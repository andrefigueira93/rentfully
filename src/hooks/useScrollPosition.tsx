import React, {
  PropsWithChildren,
  createContext,
  useContext,
  useEffect,
} from "react";

// Define the shape of the context
interface ScrollPositionContextData {
  scrollPosition: {
    x: number;
    y: number;
  };
}

// Create a context with the default value of undefined
const ScrollPositionContext = createContext<
  ScrollPositionContextData | undefined
>(undefined);

// Create a provider component that will expose the context
const ScrollPositionProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [scrollPosition, setScrollPosition] = React.useState({
    x: 0,
    y: 0,
  });

  useEffect(
    () => {
      const handleScroll = () => {
        setScrollPosition({ x: window.scrollX, y: window.scrollY });
      };

      // Add the event listener
      window.addEventListener("scroll", handleScroll);

      // Remove the event listener on cleanup
      return () => {
        window.removeEventListener("scroll", handleScroll);
      };
    },
    [
      // We don't want to run this effect on every render
      // So I'm gonna left this array of dependencies empty
      // https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects
    ]
  );

  return (
    <ScrollPositionContext.Provider value={{ scrollPosition }}>
      {children}
    </ScrollPositionContext.Provider>
  );
};

export default ScrollPositionProvider;

// Create a hook to use the ScrollPositionContext
export const useScrollPosition = () => {
  const ctx = useContext(ScrollPositionContext);
  if (!ctx)
    throw new Error(
      "`useScrollPosition` must be used within a ScrollPositionProvider"
    );
  return ctx;
};
