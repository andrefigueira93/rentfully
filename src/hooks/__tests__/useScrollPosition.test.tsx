import TestWrapper from "@/__tests__/wrapper";
import { act, renderHook } from "@testing-library/react-hooks";
import { useScrollPosition } from "../useScrollPosition";

describe("useScrollPosition hook", () => {
  test("should throw an error when used outside of ScrollPositionProvider", () => {
    const { result } = renderHook(() => useScrollPosition());

    expect(result.error).toEqual(
      Error("`useScrollPosition` must be used within a ScrollPositionProvider")
    );
  });

  test("should provide the scroll position", () => {
    const { result } = renderHook(() => useScrollPosition(), {
      wrapper: TestWrapper,
    });

    expect(result).toMatchSnapshot();
    expect(result.current.scrollPosition).toEqual({ x: 0, y: 0 });
  });

  test("should update the scroll position on scroll", async () => {
    const { result } = renderHook(() => useScrollPosition(), {
      wrapper: TestWrapper,
    });

    vi.spyOn(window, "scrollTo").mockImplementation(() => {});

    // Simulate scroll event
    act(() => {
      window.scrollTo(10, 0);
    });

    // Wait for the state to update
    expect(window.scrollTo).toHaveBeenCalledWith(10, 0);
    expect(result).toMatchSnapshot();
  });
});
