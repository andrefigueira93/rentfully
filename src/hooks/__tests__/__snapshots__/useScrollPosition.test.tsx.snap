// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useScrollPosition hook > should provide the scroll position 1`] = `
{
  "all": [
    {
      "scrollPosition": {
        "x": 0,
        "y": 0,
      },
    },
  ],
  "current": {
    "scrollPosition": {
      "x": 0,
      "y": 0,
    },
  },
  "error": undefined,
}
`;

exports[`useScrollPosition hook > should update the scroll position on scroll 1`] = `
{
  "all": [
    {
      "scrollPosition": {
        "x": 0,
        "y": 0,
      },
    },
  ],
  "current": {
    "scrollPosition": {
      "x": 0,
      "y": 0,
    },
  },
  "error": undefined,
}
`;
